### PR TITLE
Add ability to overwrite o3de object when downloading

### PR DESCRIPTION
**Pull request by AMZN-Phil**
Also check last_updated field to compare for gem updates. This method will be replaced once a gem versioning system has been introduced.

Signed-off-by: AMZN-Phil <pconroy@amazon.com>

### CODE REVIEW

File: scripts/o3de/o3de/download.py [hunk 1]
```diff
@@ -230,6 +306,9 @@ def add_parser_args(parser):
     parser.add_argument('-sar', '--skip-auto-register', action='store_true', required=False,
                             default=False,
                             help = 'Skip the automatic registration of new object download')
+    parser.add_argument('-fce', '--force-overwrite', action='store_true', required=False,
+                            default=False,
+                            help = 'Force overwrite the current object')
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/download.py:311]: I think we should go with the traditional `-f` / `--force` option the force argument.

File: scripts/o3de/o3de/utils.py [hunk 2]
```diff
@@ -117,15 +117,18 @@ def backup_folder(folder: str or pathlib.Path) -> None:
...
     :param parsed_uri: uniform resource identifier to zip file to download
     :param download_path: location path on disk to download file
     :download_progress_callback: callback called with the download progress as a percentage, returns true to request to cancel the download
     """
-    if download_path.is_file():
+    if download_path.is_file() and not force_overwrite:
         logger.warn(f'File already downloaded to {download_path}.')
-    elif parsed_uri.scheme in ['http', 'https', 'ftp', 'ftps']:
+    elif download_path.is_file():
+        shutil.rmtree(download_path)
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/utils.py:129]: This is kind strange logic, both the if and elif block are checking `download_path.is_file`.

I think an inner if block would look logically better
```suggestion
    if download_path.is_file():
        if not force_overwrite:
            logger.warn(f'File already downloaded to {download_path}.')
        else:
            shutil.rmtree(download_path)
```

File: scripts/o3de/o3de/download.py [hunk 3]
```diff
@@ -149,38 +151,108 @@ def download_o3de_object(object_name: str, default_folder_name: str, dest_path:
...
+
+    return repo_copy_updated_date > current_date
 
+def is_o3de_engine_update_available(engine_name: str, local_last_updated: str):
+    try:
+        current_updated_time = datetime.fromisoformat(local_last_updated)
+    except ValueError:
+        logger.warn(f'last_updated field has incorrect format for engine {engine_name}.')
+        # Possible that an earlier version did not have this field so still want to check against cached downloadable version
+        current_updated_time = datetime.min
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/download.py:218]: The querying of the time value from `local_last_updated` string can be done in the `is_o3de_object_update_available`

The only difference between the each implementation below is the o3de_object name.
The o3de object name itself is available via the `downloadable_kwarg_key` parameter in the `is_o3de_object_update_available` method

File: scripts/o3de/o3de/download.py [hunk 4]
```diff
@@ -149,38 +151,108 @@ def download_o3de_object(object_name: str, default_folder_name: str, dest_path:
...
-    return download_o3de_object(restricted_name, 'restricted', dest_path, 'restricted', 'restricted_name', skip_auto_register, download_progress_callback)
+    return download_o3de_object(restricted_name, 'restricted', dest_path, 'restricted', 'restricted_name', skip_auto_register, force_overwrite, download_progress_callback)
+
+def is_o3de_object_update_available(object_name: str, downloadable_kwarg_key, current_date: datetime) -> bool:
+    downloadable_object_data = get_downloadable(**{downloadable_kwarg_key : object_name})
+    if not downloadable_object_data:
+        logger.error(f'Downloadable o3de object {object_name} not found.')
+        return False
+
+    repo_copy_updated_string = 0
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/download.py:197]: This line isn't needed. The try block below will always execute.
Python variable scope isn't like C++.
As long as the line has executed, the variable exists.

File: scripts/o3de/o3de/download.py [hunk 5]
```diff
@@ -124,9 +124,11 @@ def download_o3de_object(object_name: str, default_folder_name: str, dest_path:
     if not dest_path:
         logger.error(f'Destination path cannot be empty.')
         return 1
-    if dest_path.exists():
+    if dest_path.exists() and not force_overwrite:
         logger.error(f'Destination path {dest_path} already exists.')
         return 1
+    elif dest_path.exists():
+        shutil.rmtree(dest_path)
```
Comment (AMZN-alexpete) [inline scripts/o3de/o3de/download.py:131]: one fun gotcha with rmtree is that if it fails it throws an exception, and another is that when users use Perforce it is common for folders that are checked in to be set to read-only, so rmtree() will fail here too, but you will find [workarounds for this using the onerror param](https://stackoverflow.com/questions/2656322/shutil-rmtree-fails-on-windows-with-access-is-denied).  I think we should at least catch exception and give the user a friendly log error message.  You should be able to test this by setting the gem folder to read only in Windows explorer.

Comment (AMZN-nggieber) [inline scripts/o3de/o3de/download.py:131]: Personally I think the logic flow is a little odd here I would instead check if force_overwrite is true in the scope of `if dest_path.exists()` evaluating to true. That way we only have one check if dest_path.exists().

File: scripts/o3de/o3de/download.py [hunk 6]
```diff
@@ -149,38 +151,108 @@ def download_o3de_object(object_name: str, default_folder_name: str, dest_path:
...
 
 
 
 def download_restricted(restricted_name: str,
                         dest_path: str or pathlib.Path,
                         skip_auto_register: bool,
+                        force_overwrite: bool,
                         download_progress_callback = None) -> int:
-    return download_o3de_object(restricted_name, 'restricted', dest_path, 'restricted', 'restricted_name', skip_auto_register, download_progress_callback)
+    return download_o3de_object(restricted_name, 'restricted', dest_path, 'restricted', 'restricted_name', skip_auto_register, force_overwrite, download_progress_callback)
```
Comment (AMZN-nggieber) [inline scripts/o3de/o3de/download.py:189]: I would consider splitting all of these across 2 lines. I had to horizontal scroll to see the whole call.

File: scripts/o3de/o3de/download.py [hunk 7]
```diff
@@ -149,38 +151,108 @@ def download_o3de_object(object_name: str, default_folder_name: str, dest_path:
...
+
+    try:
+        repo_copy_updated_date = datetime.fromisoformat(repo_copy_updated_string)
+    except ValueError:
+        logger.error(f'last_updated field in incorrect format for {object_name}.')
+        return False
+
+    return repo_copy_updated_date > current_date
 
+def is_o3de_engine_update_available(engine_name: str, local_last_updated: str):
```
Comment (AMZN-nggieber) [inline scripts/o3de/o3de/download.py:212]: All of these methods are really repetitive. It looks like you could create a generic method that takes the object type as a str.

File: scripts/o3de/o3de/download.py (hunk 3 above)
Comment (AMZN-alexpete) [reply scripts/o3de/o3de/download.py:218]: +1

**Review by AMZN-alexpete — COMMENTED** [review]
Some minor feedback, but otherwise is looking good.

File: scripts/o3de/o3de/download.py (hunk 3 above)
Comment (AMZN-Phil) [reply scripts/o3de/o3de/download.py:218]: Yes, good catch. Fixing this duplication also showed that the error emitted from both the local copy and the repo copy of this conversion were identical with no way to tell them apart.

File: scripts/o3de/o3de/download.py (hunk 7 above)
Comment (AMZN-nggieber) [reply scripts/o3de/o3de/download.py:258]: Nvm missed the other comment about this already.

**Review by AMZN-nggieber — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by AMZN-alexpete — APPROVED** [review]

**Pull request merged**